File: ApiWrapper/Vector3.cs
using ApiWrapper.Utils;
using System;
using System.Runtime.InteropServices;

namespace ApiWrapper
{
    public partial class Vector3 : IMarshallable
    {

        [StructLayout(LayoutKind.Sequential)]
        internal struct BackingVector
        {
            [MarshalAs(UnmanagedType.R4)]
            public float _x;
            [MarshalAs(UnmanagedType.R4)]
            public float _y;
            [MarshalAs(UnmanagedType.R4)]
            public float _z;
        }

        private IntPtr _handler;
        private BackingVector _vector;

        public Vector3()
        {
            _vector = new BackingVector();

            _handler = CreateVector3();

            Init();
        }

        public Vector3(float x, float y, float z)
        {
            _vector = new BackingVector();
            _handler = CreateVector3Args(x, y, z);
            Init();
        }

        internal Vector3(BackingVector vec)
        {
            _vector = vec;
            _handler = CreateVector3Args(vec._x, vec._y, vec._z);
        }

        public Vector3(IntPtr other)
        {
            _vector = new BackingVector();
            _handler = other;
            Init();
        }

        ~Vector3()
        {
            DeleteVector3(_handler);
        }

        internal BackingVector GetBackingVector()
        {
            return _vector;
        }


        private void Init()
        {
            _vector._x = GetX(_handler);
            _vector._y = GetY(_handler);
            _vector._z = GetZ(_handler);
        }


        public float X
        {
            get
            {
                _vector._x = GetX(_handler);
                return _vector._x;
            }
            set
            {
                SetX(_handler, value);
            }
        }
        public float Y
        {
            get
            {
                _vector._y = GetY(_handler);
                return GetY(_handler);
            }
            set
            {
                SetY(_handler, value);
            }
        }
        public float Z
        {
            get
            {
                _vector._z = GetZ(_handler);
                return GetZ(_handler);
            }
            set
            {
                SetZ(_handler, value);
            }
        }

        public object BackingField { get => _vector; set => _vector = (BackingVector) value; }

        public Type BackingFieldType => typeof(BackingVector);


        public Vector3 GetUnitVector()
        {
            return new Vector3(GetUnitVector(_handler));
        }

        public void MakeUnitVector()
        {
            MakeUnitVector(_handler);
        }

        public Vector3 Add(float amount)
        {
            return new Vector3(AddScalar(_handler, amount));
        }

        public Vector3 Add(Vector3 vec)
        {
            return new Vector3(Add(_handler, vec._handler));
        }

        public override string ToString()
        {
            return $"X : {X}, Y : {Y}, Z : {Z}";
        }
    }
}

File: ExampleAPI/Vector3.cpp
#include "Vector3.h"

Vector3C::Vector3C(float x, float y, float z)
{
	_x = x;
	_y = y;
	_z = z;
}

//Vector3C::Vector3C(const Vector3C & vec)
//{
//	_x = vec._x;
//	_y = vec._y;
//	_z = vec._z;
//}

//Vector3C::Vector3C(const Vector3 & vec)
//{
//	_vec = vec;
//}

//Vector3C::~Vector3C()
//{
//}

float Vector3C::GetX()
{
	return _x;
}

void Vector3C::SetX(float val)
{
	_x = val;
}

float Vector3C::GetY()
{
	return _y;
}

void Vector3C::SetY(float val)
{
	_y = val;
}

float Vector3C::GetZ()
{
	return _z;
}

void Vector3C::SetZ(float val)
{
	_z = val;
}

Vector3C Vector3C::GetUnitVector()
{
	float mag = Magnitude();
	return Vector3C(_x / mag, _y / mag, _z / mag);
}

void Vector3C::MakeUnitVector()
{
	float mag = Magnitude();
	_x = _x / mag;
	_y = _y / mag;
	_z = _z / mag;
}

Vector3C Vector3C::Add(Vector3C vec)
{
	return Vector3C(_x + vec._x, _y + vec._y, _z + vec._z);
}

Vector3C Vector3C::Add(float amount)
{
	return Vector3C(_x + amount, _y + amount, _z + amount);
}

float Vector3C::Magnitude()
{
	return sqrtf(_x*_x + _y * _y + _z * _z);
}

//Vector3 Vector3C::GetBackingVector()
//{
//	return _vec;
//}

MYAPI Vector3C * CreateVector3()
{
	return new Vector3C();
}

MYAPI Vector3C * CreateVector3Args(float x, float y, float z)
{
	return new Vector3C(x, y, z);
}

MYAPI void DeleteVector3(Vector3C * handler)
{
	handler->~Vector3C();
}

MYAPI float GetX(Vector3C * handler)
{
	return handler->GetX();
}

MYAPI void SetX(Vector3C * handler, float val)
{
	handler->SetX(val);
}

MYAPI float GetY(Vector3C * handler)
{
	return handler->GetY();
}

MYAPI void SetY(Vector3C * handler, float val)
{
	handler->SetY(val);
}

MYAPI float GetZ(Vector3C * handler)
{
	return handler->GetZ();
}

MYAPI void SetZ(Vector3C * handler, float val)
{
	handler->SetZ(val);
}

MYAPI Vector3C * GetUnitVector(Vector3C * handler)
{
	return new Vector3C(handler->GetUnitVector());
}

MYAPI void MakeUnitVector(Vector3C * handler)
{
	handler->MakeUnitVector();
}

MYAPI Vector3C * Add(Vector3C * handler, Vector3C * vec)
{
	return new Vector3C(handler->Add(*vec));
}

MYAPI Vector3C * AddScalar(Vector3C * handler, float amount)
{
	return new Vector3C(handler->Add(amount));
}

MYAPI float Magnitude(Vector3C * handler)
{
	return handler->Magnitude();
}
